### PR TITLE
feat(ui-shell): add product switcher component

### LINKED ISSUE #1357: Create Global Product Switcher component
This issue is for implementing the header component found in:

image

Latest sketch file: https://ibm.box.com/s/4e4zrz6mlo0jpr8lw4vimwlxn4vc59cy

### PR DESCRIPTION
**Pull request by emyarod**
Closes #1357

This PR will add functionality to the UI shell product switcher component, which extends a generic navigation panel component (name subject to change). Shares changes intended for #1359

#### Changelog

**New**

- product switcher

**Changed**

- header markup and styles to accommodate hooks needed for product switcher

### CODE REVIEW

File: src/components/ui-shell/navigation-menu-panel.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+) {
+  createdByLauncher = event => {
+    const isExpanded = !this.element.hasAttribute('hidden');
+    const newState = isExpanded ? 'collapsed' : 'expanded';
+    this.triggerButton = event.delegateTarget;
+    this.changeState(newState);
+  };
+
+  /**
+   *
```
Comment (asudoh) [inline src/components/ui-shell/navigation-menu-panel.js]: Nitpick - Do we want to update the comment here...?

File: src/globals/js/misc/on-focus-by-keyboard.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Differentiate between keyboard and mouse-triggered focusout/blur events
+ * @param {Element} node  The element to attach event listeners to
+ * @param {string} name The event name to listen to
+ * @param {Function} callback The callback function to invoke
+ * @returns {Handle} The handle to release the attached event handler
+ */
+export default function onFocusByKeyboard(node, name, callback) {
+  /**
+   *
```
Comment (asudoh) [inline src/globals/js/misc/on-focus-by-keyboard.js]: Nitpick - Do we want to update the comment here...?

**Review by asudoh — APPROVED** [review]
LGTM basically, just a couple nitpicks - Thanks @emyarod!

File: src/components/ui-shell/navigation-menu-panel.js (hunk 1 above)
Comment (emyarod) [reply src/components/ui-shell/navigation-menu-panel.js]: updated!

File: src/globals/js/misc/on-focus-by-keyboard.js (hunk 2 above)
Comment (emyarod) [reply src/globals/js/misc/on-focus-by-keyboard.js]: updated!

**Review by aledavila — APPROVED** [review]
LGTM

**Comment by aagonzales** [comment]
Which configuration is it, `switcher all products`?

**Comment by emyarod** [comment]
@aagonzales both [Default](https://deploy-preview-1528--the-carbon-components.netlify.com/component/ui-shell--default.html) and [Switcher All Products](https://deploy-preview-1528--the-carbon-components.netlify.com/component/ui-shell--switcher-all-products.html) will show the component

File: src/components/ui-shell/_header.scss [hunk 3]
```diff
@@ -29,6 +29,15 @@
     border: mini-units(0.5) solid transparent;
   }
 
+  .#{$prefix}--header__action > svg[data-navigation-menu-panel-collapse-icon],
```
Comment (joshblack) [inline src/components/ui-shell/_header.scss]: Same note as other PRs, is there any way to do this without tying the selector to the data attribute?

**Review by joshblack — CHANGES_REQUESTED** [review]

File: src/components/ui-shell/_navigation-menu.scss [hunk 4]
```diff
@@ -9,17 +9,22 @@
   // Navigation
   //----------------------------------------------------------------------------
   .#{$prefix}--navigation {
-    position: absolute;
+    position: fixed;
```
Comment (joshblack) [inline src/components/ui-shell/_navigation-menu.scss]: Are these from another PR?

**Comment by aagonzales** [comment]
I can't get it (the app switcher) to open anymore. I could open it this morning but not now ...

**Comment by emyarod** [comment]
@aagonzales it seems to be caused by a recent change in another branch, but it looks like it will be resolved when #1549 is merged into master

**Comment by emyarod** [comment]
@aagonzales it should be working again since #1549 is now resolved

**Review by aagonzales — COMMENTED** [review]
- The item hover is disappearing when you hover over option in the overflow menu. 
image
image
- Can we use the arrow icon from the new icon set for the back icon: https://carbon-elements.netlify.com/icons/examples/esm/#16%2Farrow--left

File: src/components/ui-shell/_header.scss (hunk 3 above)
Comment (emyarod) [reply src/components/ui-shell/_header.scss]: got it, should be resolved now!

**Comment by asudoh** [comment]
@emyarod Is this ready for re-review for design/dev perspectives? Thanks!

**Comment by joshblack** [comment]
_PR Reviews 1/19_

Notes

- [x] Need to fix visual review comment (icon change)

**Comment by asudoh** [comment]
@emyarod The latest tree-shaking testing result shows `!function(){\"use strict\";window}();` code, which looks fine as it doesn't contain any product switcher code. Would you want to try trimming out `window` at https://github.com/IBM/carbon-components/blob/50f3a12d653015976512a6113fa2e1932f828d21/tests/pure-modules-test.js#L110-L111? Thanks!

**Comment by emyarod** [comment]
@asudoh I believe trimming `;window` from the output should fix the failing test, although I'm not sure where `window` is being called in the product switcher code

**Comment by asudoh** [comment]
@joshblack Would you be able to re-review? Thanks!

**Review by lovemecomputer — APPROVED** [review]
lgtm!

**Review by joshblack — APPROVED** [review]

**Comment by asudoh** [comment]
OK great that everybody signed off! @emyarod could you solve the merge conflict so we can move forward with this? Thanks!

**Comment by emyarod** [comment]
should be resolved now @asudoh

**Comment by asudoh** [comment]
@emyarod Can we trim out `;window`...? Thanks!

**Comment by aagonzales** [comment]
Looks like there's a weird dotted line around the x and the blue focus is coming up on the panel as well in Chrome.

image

**Comment by emyarod** [comment]
it looks like that change was introduced in another PR (#1650) https://github.com/IBM/carbon-components/pull/1650/files#diff-33ebb84c797d2e8a44086641297274e6R182

**Comment by asudoh** [comment]
@joshblack I recall you said our vanilla repo won't need the "grid demo style" @emyarod is referring to. Is it correct?

**Comment by joshblack** [comment]
@asudoh I'm pretty sure we won't need to have it 👍 We definitely can include it if it is helpful, but not a hard requirement or anything 😄

**Comment by joshblack** [comment]
@emyarod is this ready to merge in now that the grid demo stuff is merged?

**Comment by emyarod** [comment]
yes @joshblack after @aagonzales approves I believe

**Review by aagonzales — APPROVED** [review]
Might need some more work in the future but its good for now! 🎉

**Pull request merged**